pkg/object/meshcontroller/api/api_customobject.go

```diff
@@ -0,0 +1,371 @@
+	sort.Slice(kinds, func(i, j int) bool {
+		return kinds[i].Name < kinds[j].Name
+	})
+
+	var pbKinds []*v1alpha1.CustomObjectKind
+	for _, v := range kinds {
+		kind := &v1alpha1.CustomObjectKind{}
+		err := a.convertSpecToPB(v, kind)
+		if err != nil {
+			logger.Errorf("convert spec %#v to pb spec failed: %v", v, err)
+			continue
+		}
+		pbKinds = append(pbKinds, kind)
+	}
+
+	err := json.NewEncoder(w).Encode(pbKinds)
+	if err != nil {
+		panic(fmt.Errorf("marshal %#v to json failed: %v", kinds, err))
+	}
+
+	w.Header().Set("Content-Type", "application/json")
+}
+
+func (a *API) getCustomObjectKind(w http.ResponseWriter, r *http.Request) {
+	name, err := a.readURLParam(r, "name")
+	if err != nil {
+		api.HandleAPIError(w, r, http.StatusBadRequest, err)
+		return
+	}
+
+	kind := a.service.GetCustomObjectKind(name)
+	if kind == nil {
+		api.HandleAPIError(w, r, http.StatusNotFound, fmt.Errorf("%s not found", name))
+		return
+	}
+
+	pbKind := &v1alpha1.CustomObjectKind{}
+	err = a.convertSpecToPB(kind, pbKind)
+	if err != nil {
+		panic(fmt.Errorf("convert spec %#v to pb failed: %v", kind, err))
+	}
+
+	err = json.NewEncoder(w).Encode(pbKind)
+	if err != nil {
+		panic(fmt.Errorf("marshal %#v to json failed: %v", pbKind, err))
+	}
+
+	w.Header().Set("Content-Type", "application/json")
+}
+
+func (a *API) putCustomObjectKind(w http.ResponseWriter, r *http.Request, update bool) error {
+	name, err := a.readURLParam(r, "name")
+	if err != nil {
+		api.HandleAPIError(w, r, http.StatusBadRequest, err)
+		return err
+	}
+
+	pbKind := &v1alpha1.CustomObjectKind{}
+	kind := &spec.CustomObjectKind{}
+
+	err = a.readAPISpec(r, pbKind, kind)
+	if err != nil {
+		api.HandleAPIError(w, r, http.StatusBadRequest, err)
+		return err
+	}
+	if name != kind.Name {
+		err = fmt.Errorf("name conflict: %s %s", name, kind.Name)
+		api.HandleAPIError(w, r, http.StatusConflict, err)
+		return err
+	}
+
+	if kind.JSONSchema != "" {
+		sl := gojsonschema.NewStringLoader(kind.JSONSchema)
+		if _, err = gojsonschema.NewSchema(sl); err != nil {
+			err = fmt.Errorf("invalid JSONSchema: %s", err.Error())
+			api.HandleAPIError(w, r, http.StatusBadRequest, err)
+			return err
+		}
+	}
+
+	a.service.Lock()
+	defer a.service.Unlock()
+
+	oldKind := a.service.GetCustomObjectKind(name)
+	if update && (oldKind == nil) {
+		err = fmt.Errorf("%s not found", name)
+		api.HandleAPIError(w, r, http.StatusNotFound, err)
+		return err
+	}
+	if (!update) && (oldKind != nil) {
+		err = fmt.Errorf("%s existed", name)
+		api.HandleAPIError(w, r, http.StatusConflict, err)
+		return err
+	}
+
+	a.service.PutCustomObjectKind(kind)
+	return nil
+}
+
+func (a *API) createCustomObjectKind(w http.ResponseWriter, r *http.Request) {
+	err := a.putCustomObjectKind(w, r, false)
+	if err == nil {
+		w.Header().Set("Location", r.URL.Path)
+		w.WriteHeader(http.StatusCreated)
+	}
+}
+
+func (a *API) updateCustomObjectKind(w http.ResponseWriter, r *http.Request) {
```

Comment:
`updateCustomObjectKind` is a handler from exposed HTTP API, I think using `put` in the function name is more semantics consistent. And `update`, `create`, or `delete` is more implementation storage related, I prefer to use them in the inner structure's function. 
Anyway, this is a suggestion that is not a very strong request, it's up to u.

Reply:
updated